File: Turash/Turash_lab4/source/turash_lab4(csharp)/turash_lab4/GAIndividual.cs
using System;

namespace turash_lab4
{
    class GAIndividual
    {
        private static Random randg = new Random(); // Рандомний генератор
        public int GenomeSize;
        public double[] Genome;
        public double Fitness;

        public GAIndividual(int gsize, double[] minRange, double[] maxRange)
        {
            // створити випадкового індивідума довжиною gsize
            // і-й ген повине бути в діапазоні між minRange[i] і maxRange[i]
            GenomeSize = gsize;
            Genome = new double[GenomeSize];
            for (var i = 0; i < GenomeSize; i++)
            {
                Genome[i] = (double)randg.NextDouble() * (maxRange[i] - minRange[i]) + minRange[i];
            }


            EvalFitness();// оцінити придатність цього нового індивідума
        }

        public GAIndividual(double[] d)
        {
            // Створити індивідума, що його ген є таки же як d []

            GenomeSize = d.Length;
            Genome = new double[GenomeSize];
            for (var i = 0; i < GenomeSize; i++)
            {
                Genome[i] = d[i];
            }
            EvalFitness();// оцінити придатність цієї нового індивідума
        }

        public GAIndividual Mutate(double[] minRange, double[] maxRange)
        {
            // rate це шанс кожного гена мутувати
            var rate = 1.0f / (double)GenomeSize;
            var result = new double[GenomeSize];
            for (var i = 0; i < GenomeSize; i++)
                result[i] = Genome[i];
            // застосування точкової мутації
            for (var i = 0; i < GenomeSize; i++)
                result[i] = randg.NextDouble() * (maxRange[i] - minRange[i]) + minRange[i];

            return new GAIndividual(result);
        }

        public static GAIndividual Xover1P(GAIndividual f, GAIndividual m)
        {
            // одноточковий кросинговер
            var rng = new Random();
            int xpoint = 1 + rng.Next(1);
            double[] child = new double[f.GenomeSize];
            for (var i = 0; i < xpoint; i++)
            {
                child[i] = f.Genome[i];
            }
            for (int i = xpoint; i < f.GenomeSize; i++)
            {
                child[i] = m.Genome[i];
            }
            return new GAIndividual(child);
        }

        public override string ToString()
        {
            var s = "[";
            s += Math.Round(Genome[0], 5) + "]";
            s += " Fitness = " + Math.Round(Fitness, 5);
            return s;
        }

        private void EvalFitness() => Fitness = -(62 - Genome[0] - 86 * Genome[0] * Genome[0] + 2 * Genome[0] * Genome[0] * Genome[0]);
    }
}

File: Yerchenko/Yerchenko_lab4/GAFunc/src/GAIndividual.java
import java.util.Random;

public class GAIndividual {
	private static Random randg = new Random(); // random gen.
	public int genome_size;
	public float[] genome;
	public float fitness;

	public GAIndividual(int gsize, float[] min_range, float[] max_range) {
		// random individual with length gsize
		// i-gen must be in range between min_range[i] and max_range[i]
		genome_size = gsize;
		genome = new float[genome_size];
		for (int i = 0; i < genome_size; i++) {
			genome[i] = randg.nextFloat() * (max_range[i] - min_range[i])
					+ min_range[i];
		}
		///////////
					//System.out.println(genome[0]);
		evalFitness();//evaluate the new individual
	}

	public GAIndividual(float d[]) {
		// create individual with same gene as d[]
		// min_range[i] <= d[i] <= max_range[i]
		genome_size = d.length;
		genome = new float[genome_size];
		for (int i = 0; i < genome_size; i++) {
			genome[i] = d[i];
		}
		evalFitness();//evaluate the new individual
	}

	public GAIndividual mutate(float[] min_range, float[] max_range) {
		float rate = 1.0f / (float) genome_size;
		float[] result = new float[genome_size];
		for (int i = 0; i < genome_size; i++)
			result[i] = genome[i];
		// point mutation
		for (int i = 0; i < genome_size; i++)
			if ((float) Math.random() < rate)
				result[i] = randg.nextFloat() * (max_range[i] - min_range[i])
						+ min_range[i];

		return new GAIndividual(result);
	}

	public static GAIndividual xover1p(GAIndividual f, GAIndividual m) {
		// single point crossover
		Random rng = new Random();
		int xpoint = 1 + rng.nextInt(1);
		float[] child = new float[f.genome_size];
		for (int i = 0; i < xpoint; i++) {
			child[i] = f.genome[i];
		}
		for (int i = xpoint; i < f.genome_size; i++) {
			child[i] = m.genome[i];
		}
		return new GAIndividual(child);
	}

	public String toString() {
		String s = "[";
		s += genome[0] + "]";
		s += " fitness = " + fitness;
		return s;
	}

	private void evalFitness() {

		fitness = (39 - 96 * genome[0] - 67 * genome[0] * genome[0] + 4 * genome[0]
						* genome[0] * genome[0]);

	}

}
